src/lsm/block_free_set.zig

```diff
@@ -1,37 +1,421 @@
 const std = @import("std");
 const assert = std.debug.assert;
 const mem = std.mem;
+const DynamicBitSetUnmanaged = std.bit_set.DynamicBitSetUnmanaged;
+const MaskInt = DynamicBitSetUnmanaged.MaskInt;
+const config = @import("../config.zig");
+const RingBuffer = @import("../ring_buffer.zig").RingBuffer(
+    usize,
+    config.cache_line_size / @sizeOf(usize), // 1 cache line of recently-freed blocks
+    .array,
+);
+const EWAH = @import("../ewah.zig").EWAH(usize);
 
-/// The 0 address is reserved for usage as a sentinel and will never be returned
-/// by acquire().
-/// TODO add encode/decode function for run-length encoding to/from SuperBlock.
+/// The 0 address is reserved for usage as a sentinel and will never be returned by acquire().
+///
+/// Set bits indicate free blocks, unset bits are allocated.
 pub const BlockFreeSet = struct {
-    /// Bits set indicate free blocks
-    free: std.bit_set.DynamicBitSetUnmanaged,
+    // Each bit of `index` is the OR of `shard_size` bits of `blocks`.
+    // That is, if a shard has any free blocks, the corresponding index bit is set.
+    index: DynamicBitSetUnmanaged,
+    blocks: DynamicBitSetUnmanaged,
+    // A fast cache of the 0-indexed bits (not 1-indexed addresses) of recently freed blocks.
+    recent: RingBuffer,
+
+    // Fixing the shard size to a constant rather than varying the shard size (but
+    // guaranteeing the index always a multiple of 64B) means that the top-level index
+    // may have some unused bits. But the shards themselves are always a multiple of
```

Comment:
Is this comment still in sync with our idea to put more constraints on the number of blocks managed by `BlockFreeSet`? It would be great to restrict the cardinality of `blocks_count` in such a way as to avoid any padding like this if possible.

Reply:
Afaik this comment is still correct. During my first pass, I tried it the other way around (i.e. make the shard size more flexible to guarantee no padding) but it's much messier.

---

> It would be great to restrict the cardinality of `blocks_count` in such a way as to avoid any padding like this if possible.

Avoiding padding is just a matter of configuring a `blocks_count` which is a multiple of

```
word_size × shard_size = 64 * 4096 = 262_144
```

In other words, ensure that the number of shards (and thus index bits) is a multiple of the word size.

Reply:
We should be able to drop the `means that the top-level index may have some unused bits` comment since we're now requiring it to be exact?